### PR TITLE
Add protection for mention spam

### LINKED ISSUE #309: Mention flood protection
As https://github.com/matrix-org/mjolnir/pull/210 has been closed, I want to make a feature request here to at least track this, as i'd like to see this in mjolnir.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #309 #211 

This protection checks whether a message contains a large number of mentions and redacts the message. It currently checks the mentions array, then the body and finally the formatted body. I've chosen 10 as a sensible max number of mentions before the message may be considered spam, although this is configurable via a setting.

### CODE REVIEW

File: src/protections/MentionSpam.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+limitations under the License.
+*/
+
+import { Protection } from "./IProtection";
+import { Mjolnir } from "../Mjolnir";
+import { LogLevel, Permalinks, UserID } from "@vector-im/matrix-bot-sdk";
+import { NumberProtectionSetting } from "./ProtectionSettings";
+
+export const DEFAULT_MAX_MENTIONS = 10;
+const USER_ID_REGEX = /@[^:]*:.+/;
```
Comment (turt2live) [inline src/protections/MentionSpam.ts:23]: can we just limit this to number of `@` symbols? (plus HTML encoded varient for htmlBody)

**Review by turt2live — APPROVED** [review]
generally lgtm

File: src/protections/MentionSpam.ts (hunk 1 above)
Comment (davidegirardi) [reply src/protections/MentionSpam.ts:23]: I think there are concrete chances of that reacting on somebody posting a snippet of code with some delimiter:
```ini
some = value
;@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
next = section
```

Comment (turt2live) [reply src/protections/MentionSpam.ts:23]: "don't do that in my rooms" :p

File: src/protections/MentionSpam.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    public checkMentions(body: unknown|undefined, htmlBody: unknown|undefined, mentionArray: unknown|undefined): boolean {
+        const max = this.settings.maxMentions.value;
+        if (Array.isArray(mentionArray) && mentionArray.length > max) {
+            return true;
+        }
+        if (typeof body === "string" && body.split('@').length - 1 > max) {
+            return true;
+        }
+        if (typeof htmlBody === "string" && htmlBody.split('%40').length - 1 > max) {
```
Comment (Gnuxie) [inline src/protections/MentionSpam.ts:49]: I don't think all clients urlencode mxids properly in the anchor hrefs. So just a heads up that this isn't perfect

Comment (MTRNord) [reply src/protections/MentionSpam.ts:49]: (famously element didnt in the past. Not sure if this has been fixed. I just remember other client devs implementing workarounds for this case)

Comment (Half-Shot) [reply src/protections/MentionSpam.ts:49]: why even have standards :fire: 

thanks, I'll address this one when I'm back if someone doesn't get to it first.

**Pull request merged**